media_softlet/linux/common/os/i915/mos_bufmgr.c

```diff
@@ -2964,6 +2965,79 @@ mos_update_buffer_offsets2 (struct mos_bufmgr_gem *bufmgr_gem, mos_linux_context
     }
 }
 
+//todo: to move synchronization_xe.h to os common instead of xe specific
+#include "mos_synchronization_xe.h"
+
+int
+__add_eb_fence_array(struct mos_bufmgr_gem *bufmgr_gem,
+            struct drm_i915_gem_execbuffer2 *eb,
+            unsigned int flags)
+{
+#define SCALABILITY_ON (I915_EXEC_FENCE_OUT | I915_EXEC_FENCE_IN | I915_EXEC_FENCE_SUBMIT)
+
+    //Ignore multi batch submission for scalability to simplify logic
+    //todo: check has_fence_array from params
+    if (!(flags & SCALABILITY_ON)
+                && bufmgr_gem->exec_fences.fences)
+    {
+        int32_t fence_count = bufmgr_gem->exec_fences.count;
+        int32_t *exec_fences = bufmgr_gem->exec_fences.fences;
+
+        if (fence_count > 0)
+        {
+            struct drm_i915_gem_exec_fence *fences
+                = (struct drm_i915_gem_exec_fence *)malloc(fence_count * sizeof(struct drm_i915_gem_exec_fence));
+            if (fences == nullptr)
+            {
+                return -ENOMEM;
+            }
+            for (int32_t i = 0; i < fence_count; i++)
+            {
+                fences[i].handle = mos_sync_syncfile_fd_to_syncobj_handle(bufmgr_gem->fd, exec_fences[i + 1]);
```

Comment:
to check if exec_fences[i + 1] is valid and imported status; Ignore invalid one.